webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,760 @@
+    panelCloudNotification = document.getElementById("panelCloudNotification");
+    panelCloudNotification.removeAttribute("hidden");
+    panelCloudNotification.addEventListener("click", this);
+  },
+
+  _addRadioOption(key, providerName, document) {
+    const option = document.createElement("radio");
+    option.id = key;
+    option.type = "radio";
+
+    option.setAttribute("class", "panelCloudNotificationUI-radio");
+    option.setAttribute("label", "Save to " + providerName);
+    option.setAttribute("provider", this._formatProviderName(providerName));
+    return option;
+  },
+
+  _addNotificationMultipleProviders(providersMap, document) {
+    const multiProviderSelect = document.getElementById("multiProviderSelect");
+
+    const providers = [];
+    providersMap.forEach((value, key) => {
+      providers.push({"key": key, "name": value.displayName});
+    });
+
+    // Reset multiProviderSelect by removing any previous
+    // added options
+    while (multiProviderSelect.firstChild) {
+      multiProviderSelect.firstChild.remove();
+    }
+
+    for (const provider of providers) {
+      multiProviderSelect.appendChild(
+        this._addRadioOption(provider.key, provider.name, document));
+    }
+    multiProviderSelect.removeAttribute("hidden");
+  },
+
+  _iconURL(name) {
+    if (this._formatProviderName(name) === "localdownload") {
+      return "moz-icon://" + CloudDownloadsInternal.defaultDownloadDirIconURL + "?size=16";
+    }
+    return new URL(this.stylesURL).origin + "/skin/" + this._formatProviderName(name) + ".svg";
+  },
+
+  _setMenuItemAttributes(menuItem, providerKey, providerDetail, downloadElement) {
+    menuItem.setAttribute("label", providerDetail.displayName);
+    menuItem.setAttribute("providerKey", providerKey);
+    menuItem.setAttribute("class", "menuitem-iconic");
+    menuItem.setAttribute("image", this._iconURL(providerDetail.displayName));
+    return menuItem;
+  },
+
+  _setMoveDownloadMenuPopUpAttributes(menuItem) {
+    menuItem.setAttribute("label", "Move Download To...");
+    menuItem.removeAttribute("hidden");
+    const menuPopup = menuItem.menupopup;
+    menuPopup.removeAttribute("hidden");
+    return menuPopup.firstChild;
+  },
+
+  _displayMoveToCloudContextMenuItem(downloadElement) {
+    const aPopupMenu = downloadElement.ownerDocument.getElementById("downloadsContextMenu");
+    let menuItem = aPopupMenu.getElementsByAttribute("id", "moveDownload")[0];
+    if (menuItem) {
+      menuItem.setAttribute("hidden", "true");
+
+      // Exit if user has opted to store subsequent downloads in cloud provider folder
+      if (CloudDownloadsInternal.preferredProviderKey) {
+        return;
+      }
+
+      if (CloudDownloadsInternal.checkIfExistingCloudProviderDownloadSettings()) {
+        return;
+      }
+
+      const downloadType = downloadElement.getAttribute("cloudstorage");
+      // Downloaded Item is saved in local download folder
+      if (downloadType === "local") {
```

Comment:
The contents of the `if` and `else` blocks here are almost identical. It looks like they could reasonably be unified. Something like:

```js
        if (this.providers.size > 1) {
          let subMenuItem =  this._setMoveDownloadMenuPopUpAttributes(menuItem);
          this.providers.forEach((value, key) => {
            if (this._formatProviderName(value.displayName) !== downloadType) {
              subMenuItem = this._setMenuItemAttributes(subMenuItem, key, value, downloadElement);
              subMenuItem = subMenuItem.nextSibling;
            }
          });
          if (downloadType != "local") {
            subMenuItem = this._setMenuItemAttributes(subMenuItem, "local", {displayName: "Local Download"}, downloadElement);
          }
        } else {
          const provider = downloadType != "local" ?
            ["local", {displayName: "Local Download"}] : this.providers.entries().next().value;
          menuItem = this._setMenuItemAttributes(menuItem, provider[0], provider[1], downloadElement);
          menuItem.setAttribute("label", "Move To " + provider[1].displayName);
          menuItem.removeAttribute("hidden");
        }
```

Reply:
definitely more concise, thank you!